cosmovisor/scanner.go

```diff
@@ -1,42 +1,133 @@
 package cosmovisor
 
 import (
-	"bufio"
-	"regexp"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"time"
 )
 
-// Trim off whitespace around the info - match least greedy, grab as much space on both sides
-// Defined here: https://github.com/cosmos/cosmos-sdk/blob/release/v0.38.2/x/upgrade/abci.go#L38
-//  fmt.Sprintf("UPGRADE \"%s\" NEEDED at %s: %s", plan.Name, plan.DueAt(), plan.Info)
-// DueAt defined here: https://github.com/cosmos/cosmos-sdk/blob/release/v0.38.2/x/upgrade/internal/types/plan.go#L73-L78
-//
-//    if !p.Time.IsZero() {
-//      return fmt.Sprintf("time: %s", p.Time.UTC().Format(time.RFC3339))
-//    }
-//    return fmt.Sprintf("height: %d", p.Height)
-var upgradeRegex = regexp.MustCompile(`UPGRADE "(.*)" NEEDED at ((height): (\d+)|(time): (\S+)):\s+(\S*)`)
-
-// UpgradeInfo is the details from the regexp
+// UpgradeInfo is the update details created by `x/upgrade/keeper.DumpUpgradeInfoToDisk`.
 type UpgradeInfo struct {
-	Name string
-	Info string
+	Name   string `json:"name"`
+	Info   string `json:"info"`
+	Height uint   `json:"height"`
 }
 
-// WaitForUpdate will listen to the scanner until a line matches upgradeRegexp.
-// It returns (info, nil) on a matching line
-// It returns (nil, err) if the input stream errored
-// It returns (nil, nil) if the input closed without ever matching the regexp
-func WaitForUpdate(scanner *bufio.Scanner) (*UpgradeInfo, error) {
-	for scanner.Scan() {
-		line := scanner.Text()
-		if upgradeRegex.MatchString(line) {
-			subs := upgradeRegex.FindStringSubmatch(line)
-			info := UpgradeInfo{
-				Name: subs[1],
-				Info: subs[7],
+type fileWatcher struct {
+	// full path to a watched file
+	filename string
+	interval time.Duration
+
+	currentInfo UpgradeInfo
+	lastModTime time.Time
+	cancel      chan bool
+	ticker      *time.Ticker
+	needsUpdate bool
+
+	initialized bool
+}
+
+func newUpgradeFileWatcher(filename string, interval time.Duration) (*fileWatcher, error) {
+	if filename == "" {
+		return nil, errors.New("filename undefined")
+	}
```

Comment:
Should we check if the file exists rather than checking if its non-empty? It seems strange that we check and error if the path is invalid or the directory does not exist but we do not check if the file exists. At the moment, someone can provide the wrong filename to `DAEMON_UPGRADE_INFO_FILE` and cosmovisor will start without error.